### ISSUE TITLE
handleRedirectPromise flakily returns null after redirect in Microsoft Teams authentication window

### ISSUE THREAD

**Issue by benlieber-codaio** [AzureAD/microsoft-authentication-library-for-js#4658]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.22.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.3.0

### Background

Using msal-react within a Microsoft Teams tab configuration page (effectively a single-page-app) to get graph access to read from channel/chat members since Microsoft Teams authentication libraries don't seem provide a way to do this otherwise (token they give you only has a limited set of scopes). Originally used `loginPopup` which worked consistently in the browser but isn't compatible in the teams desktop client as this popup is blocked. One solution found to work around this is to use the microsoft teams authentication library to open an authentication window (this kind of popup is allowed) and then within this authentication window to redirect. My code is roughly inspired from this project from ~2 years ago which does this https://github.com/nmetulev/teams-msal

### Issue

The authentication window frequently enters redirect loops as in it's not able to detect the handshake back from a prior redirect so it attempts to redirect again. In particular `handleRedirectPromise` returns null, and calling `getAllAccounts` returns no accounts. However, occasionally (maybe 1 out of 10-20 times) this will succeed.

### Potential Related Issues

Closest related issue could find https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4444 , though suspect I'm running into a different issue as in my case this repros in any browser (teams desktop client, chrome, firefox) and succeeds sometimes.

### Error Message

No error is thrown, handleRedirectPromise just returns null.

### Msal Logs

![image](https://user-images.githubusercontent.com/76975174/161355582-87466254-743c-4997-87e1-d2914091263f.png)

### MSAL Configuration

```javascript
...
  {
    auth: {
      clientId: config.microsoftTeams.clientId,
      redirectUri,
    },
    cache: {
      // Note(benlieber) - use local storage as want to be able to share across redirect auth window and main config tab
      // See https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/caching.md
      cacheLocation: 'localStorage',
      storeAuthStateInCookie: false,
    },
  ...
```

### Relevant Code Snippets

```javascript
// Note(benlieber) - currently only have 1 set of scopes tab config uses so just hard-coding to avoid having to plumb
// through redirect workflow
const scopes = ListMembersScopes as string[];

function GraphAccessAuthHandlerComponentImpl() {
  const {context} = useMicrosoftTeams();
  const {instance} = useMsal();

  useEffect(() => {
    if (!context) {
      return;
    }

    instance.addEventCallback((message: {eventType: EventType; error: Error; payload: {accessToken: string}}) => {
      switch (message.eventType) {
        case EventType.LOGIN_FAILURE:
        case EventType.ACQUIRE_TOKEN_FAILURE:
          microsoftTeams.authentication.notifyFailure(message.error.message);
        // TODO(benlieber) - this is "just in case" code, it should be redundant with `handleRedirectPromise` however
        // this setup is generally flaky
        case EventType.LOGIN_SUCCESS:
        case EventType.ACQUIRE_TOKEN_SUCCESS:
          microsoftTeams.authentication.notifySuccess(message.payload.accessToken);
      }
    });

    const redirect = async () => {
      // If we are previously redirected handle this instead of redirecting again.
      const result = await instance.handleRedirectPromise();
      if (result) {
        microsoftTeams.authentication.notifySuccess(result.accessToken);
        return;
      }

      const account = instance.getActiveAccount();
      if (account) {
        await instance.acquireTokenRedirect({scopes, account});
      } else {
        await instance.loginRedirect({scopes});
      }
    };
    void redirect();
  }, [context, instance]);

  return <LoadingIndicator className={LoadingIndicator.style.medium} />;
}
const GraphAccessAuthHandlerComponent = functionalComponent(
  'GraphAccessAuthHandlerComponent',
  GraphAccessAuthHandlerComponentImpl,
);

function GraphAccessAuthHandlerImpl() {
  return (
    <MicrosoftTeamsMsalProvider redirectUri={`${config.host}/msteams/graphAuth`}>
      <GraphAccessAuthHandlerComponent />
    </MicrosoftTeamsMsalProvider>
  );
}
export const GraphAccessAuthHandler = functionalComponent('GraphAccessAuthHandler', GraphAccessAuthHandlerImpl);

/** Get a graph access token to query Microsoft Graph APIs in the browser on behalf of the user.
 *  This is independent of all Coda authentication and login logic, it's solely for the purpose of being able to query
 *  additional data in the client from Microsoft.
 *
 *  Note that the application must also be wrapped in MicrosoftTeamsMsalProvider for this to work properly.
 *
 *  See https://docs.microsoft.com/en-us/azure/active-directory/develop/tutorial-v2-react
 *  and also https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/auth-silent-aad
 */
export function useGetGraphAccessToken() {
  const {instance, accounts} = useMsal();

  const getGraphAccessToken = useCallback(async (): Promise<string> => {
    if (accounts.length > 1) {
      throw new Error('Resolved multiple accounts for user, cannot determine which one to use');
    }

    if (accounts.length === 1) {
      const [account] = accounts;
      try {
        // Attempt to silently acquire an access token which is then attached to a request for Microsoft Graph data.
        const {accessToken} = await instance.acquireTokenSilent({scopes, account});
        return accessToken;
      } catch {
        // If this fails also kick open a popup to request auth from the user.
      }
    }

    // Pop open a window to allow user to sign in + grant permissions for scopes.
    // Unfortunately msal-react 'popup' methods are blocked in teams, so as a workaround have to kick open an auth
    // window using provided Microsoft teams hook, redirect within this auth window, then pass back the access token.
    const accessToken = await teamsAuthenticate(`${config.host}/msteams/graphAuth`);
    return ensureExists(accessToken);
  }, [accounts, instance]);

  return getGraphAccessToken;
}

export async function teamsAuthenticate(url: string): Promise<string | undefined> {
  return new Promise<string | undefined>((resolve, reject) => {
    microsoftTeams.authentication.authenticate({
      url,
      successCallback: result => resolve(result),
      failureCallback: error => reject(error),
    });
  });
}

function getMsalConfig(redirectUri: string) {
  return {
    auth: {
      clientId: config.microsoftTeams.clientId,
      redirectUri,
    },
    cache: {
      // Note(benlieber) - use local storage as want to be able to share across redirect auth window and main config tab
      // See https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/caching.md
      cacheLocation: 'localStorage',
      storeAuthStateInCookie: false,
    },
    system: {
      loggerOptions: {
        loggerCallback: (level: LogLevel, message: string, containsPii: boolean): void => {
          if (containsPii) {
            return;
          }
          switch (level) {
            case LogLevel.Error:
              log.error(message);
              return;
            case LogLevel.Info:
              log.info(message);
              return;
            case LogLevel.Verbose:
              // TODO(benlieber) - make this log.verbose once this is stable
              log.info(message);
              return;
            case LogLevel.Warning:
              log.warn(message);
              return;
          }
        },
      },
    },
  };
}

function MicrosoftTeamsMsalProviderImpl({redirectUri, children}: PropsWithChildren<{redirectUri: string}>) {
  const msalInstance = new PublicClientApplication(getMsalConfig(redirectUri));
  return <MsalProvider instance={msalInstance}>{children}</MsalProvider>;
}
export const MicrosoftTeamsMsalProvider = functionalComponent(
  'MicrosoftTeamsMsalProvider',
  MicrosoftTeamsMsalProviderImpl,
);

export function useMicrosoftTeams() {
  const [initialized, setInitialized] = useState<boolean>(false);
  const [context, setContext] = useState<microsoftTeams.Context>();

  useMountEffect(() => {
    microsoftTeams.initialize(() => {
      microsoftTeams.getContext(ctx => {
        setInitialized(true);
        setContext(ctx);
      });
    });
  });

  ...
}
```

### Reproduction Steps

1. Run the following code in a [Microsoft Teams tab configuration page](https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/create-tab-pages/configuration-page)
2. Observe that sometimes the opened authentication window will sometimes enter a redirect loop (and sometimes successfully redirect and return an access token)

### Expected Behavior

handleRedirectPromise should consistently return an `AuthenticationResult` or throw an error if login failed respectively after handshaking back from `loginRedirect` or `acquireTokenRedirect`

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome, Firefox, Other

### Source

External (Customer)

**Comment by sameerag** [comment]
@tnorling Have we seen this before? msal-react used in a teams tab (embedded app as a SPA) has issues with `redirect` and the `popup` is disabled by teams.

**Comment by tnorling** [comment]
@benlieber-codaio Please check out [this sample](https://github.com/pnp/teams-dev-samples/tree/main/samples/tab-sso/src/nodejs) which demonstrates our recommended approach for integrating with Teams and let us know if you have any follow up questions.

**Comment by benlieber-codaio** [comment]
@tnorling thanks for the pointer - took a look through the sample and would still appreciate some further help diagnosing the specific issue reported here.

Some notes comparing the two - it seems like the pattern for handling the redirect is similar in that it uses the event callback `instance.addEventCallback` https://github.com/pnp/teams-dev-samples/blob/main/samples/tab-sso/src/nodejs/src/components/ClosePopup.js#L27 but interestingly enough doesn't call `handleRedirectPromise`. Another difference is that `navigateToLoginRequestUrl` is set to false https://github.com/pnp/teams-dev-samples/blob/main/samples/tab-sso/src/nodejs/src/components/App.js#L30 and there is a different handler for auth-end v.s. auth-begin - I would imagine either configuration should be workable though.

Another hypothesis could be if auth is actually failing under the hood unrelated to the redirect construction and the SDK just isn't propagating the error correctly.

**Comment by jasonnutter** [comment]
> but interestingly enough doesn't call handleRedirectPromise

With MSAL React, applications do not need to manually invoke `handleRedirectPromise`. More details here: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/FAQ.md#how-do-i-handle-the-redirect-flow-in-a-react-app

The first thing I would look at is to ensure the page you are using for the redirect URI for loginRedirect​ inside the Teams dialog is not clearing the fragment/hash from the url before MSAL has a chance to process it. This is a common issue when using frameworks like Angular and React. It looks like you are using React, are you using a router library (e.g. React Router)? Furthermore, I would recommend debugging handleRedirectPromise​ and see why MSAL doesn't believe there is an interaction in progress that it should be handling (which is why it would return null​ by design).

**Comment by benlieber-codaio** [comment]
@jasonnutter I tried updating to match the pattern employed in the sample using the event callback and a different handler for auth end - still ran into the same issue where sometimes the auth end will just hang i.e. the event is never received (fwiw the documentation link you provided also does say you can use `handleRedirectPromise` )
<img width="1078" alt="image" src="https://user-images.githubusercontent.com/76975174/161863156-19bc3a27-bca0-410c-bc0b-1bc0958e8698.png"> 

> The first thing I would look at is to ensure the page you are using for the redirect URI for loginRedirect​ inside the Teams dialog is not clearing the fragment/hash from the url before MSAL has a chance to process it. This is a common issue when using frameworks like Angular and React. It looks like you are using React, are you using a router library (e.g. React Router)?

this feels like a good tip, appreciate the pointer. we have a somewhat custom router framework that also uses React Router -  still digging into this but don't yet have evidence to suggest that we would be redirecting/clearing the hash code - have you seen this or reports accordingly from using React Router?

it's a bit tricky to get a reliable repo as the issue is transient (and attaching a debugger seems to affect the timing), setting a break point on redirects / hash changes I was able to catch `handleRedirectPromise` itself clearing the hash code but nothing on our end (thus far at least). also not sure if this `handleRedirectPromise` is the one we are calling or the one that's called within the `msal-react` lib

![image](https://user-images.githubusercontent.com/76975174/161864114-b26c4398-db19-4cc2-b2a4-deb9f19269db.png)

coming from https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/2e26302a8c70a93a094e62f1181b1029596ae7c6/lib/msal-browser/src/interaction_client/RedirectClient.ts#L95

> Furthermore, I would recommend debugging handleRedirectPromise​ and see why MSAL doesn't believe there is an interaction in progress that it should be handling (which is why it would return null​ by design).

So far I haven't seen particularly descriptive logs here (some attached in initial report), which is in large part why reporting this issue and was hoping for some expert help in the area.

**Comment by benlieber-codaio** [comment]
some more logs was able to grab - this time using the event handlers only (in my code) and separating out the auth start and auth end endpoints. note the flow observed in terms of urls was essentially `authStart` -> `authEnd#code=...` -> `authEnd` (hangs at `authEnd` as event is never received)

pre-redirect (from another attempt but assuming the same):

![image](https://user-images.githubusercontent.com/76975174/161866128-891c3579-3d3f-4fee-a389-b7f56b0cf2ff.png)

post-redirect:

<img width="992" alt="image" src="https://user-images.githubusercontent.com/76975174/161865979-4f3a768a-1685-4e47-bc9e-4777cfdf93a5.png">

**Comment by jasonnutter** [comment]
@benlieber-codaio Thanks for the addition info.

> this feels like a good tip, appreciate the pointer. we have a somewhat custom router framework that also uses React Router - still digging into this but don't yet have evidence to suggest that we would be redirecting/clearing the hash code - have you seen this or reports accordingly from using React Router?

I believe we have, yes. One thing you can try is to set the page in the dialogue (to trigger loginRedirect and use as the redirectURI) to a completely blank HTML page outside of your React (i.e. https://myapplication/msalredirect.html) that just runs the code to trigger loginRedirect, handleRedirectPromise, the logic to communicate back to your app in Teams. This can help eliminate your router (or something else in your React app) as the cause.

**Comment by benlieber-codaio** [comment]
@jasonnutter so i did the opposite of your suggestion but i _think_ should help to isolate about the same - stripped out all the msal related code from my auth end handler and only observe `authStart` -> `authEnd#code=...` . so it seems like the (only) actor that's clearing the hash code is the msal lib itself

**Comment by benlieber-codaio** [comment]
one more hypothesis - suspicious the issue might actually be a race condition in that I'm not immediately listening for events / calling `handleRedirectPromise` in my redirect handler. instead wait for both (1) microsoft teams lib to be initialized (2) msal instance to be initialized. given that msal-react calls `handleRedirectPromise` [under the hood ](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/e6b3968d047ce6e3bdb222223ae4282eb18227e3/lib/msal-react/src/MsalProvider.tsx#L120) could be an issue that this runs and clears the hash code before my code tries to do the same in the case that the microsoft teams lib takes longer to initialize than msal

(this code in particular)
```
function GraphAccessAuthHandlerComponentImpl() {
  const {context} = useMicrosoftTeams();
  const {instance} = useMsal();

  useEffect(() => {
    if (!context) {
      return;
    }

    instance.addEventCallback((message: {eventType: EventType; error: Error; payload: {accessToken: string}}) => {
    ...
```

**Comment by jasonnutter** [comment]
@benlieber-codaio Ah, that could be it. Given you are using `useMsal`, I would recommend just relying on the event API to get the result (and [inspect the interaction status](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/hooks.md#usemsal-hook) instead of waiting on `handleRedirectPromise`). And furthermore, it may also help to use two dedicated pages in the Teams dialog, one that invokes login, and one that process the login result (as the linked sample shows), which will probably help avoid loops.

@tnorling Any other thoughts?

**Comment by benlieber-codaio** [comment]
> I would recommend just relying on the event API to get the result (and [inspect the interaction status](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-react/docs/hooks.md?rgh-link-date=2022-04-06T17%3A15%3A10Z#usemsal-hook) instead of waiting on handleRedirectPromise). And furthermore, it may also help to use two dedicated pages in the Teams dialog, one that invokes login, and one that process the login result (as the linked sample shows), which will probably help avoid loops.

@jasonnutter @tnorling I updated my code to do this and updated the hook to avoid the race condition with teams initialization. now running into another issue, seeing error message `state_not_found: State not found:  Cached State` . guessing might indicate an issue where msal-react under the hood is still failing to update the cache / accounts overall?

**Comment by jasonnutter** [comment]
@benlieber-codaio Does this error go away if you enable `storeAuthStateInCookie`?

**Comment by benlieber-codaio** [comment]
> @benlieber-codaio Does this error go away if you enable `storeAuthStateInCookie`?

@jasonnutter seem to still have the same error if set `storeAuthStateInCookie` to true

**Comment by benlieber-codaio** [comment]
another debugging note - the issue is similarly transient - sometimes it will succeed.

this post suggested needed to also include removing the event callback in the return from `useEffect` which I hadn't done - is this the case? http://ostack.cn/qa/?qa=108886/

**Comment by jasonnutter** [comment]
@benlieber-codaio Interesting, thanks. Sounds like a good idea, worth trying. 

When this happens, what is in session storage? Is it empty? Or are there entries (the state value MSAL is looking for is stored in session storage [and cookies if `storeAuthStateInCookie` is enabled])?

**Comment by benlieber-codaio** [comment]
@jasonnutter I ended up going with using just `msal-browser` and calling `handleRedirectPromise` directly for the end auth handler. Haven't had a ton of time to bake it yet but feels more stable so far / haven't run into any flakiness yet.

If it helps to diagnose further attaching code snippets - new code using `msal-browser` directly:

```
// Note(benlieber) - use msal-browser instead of msal-react due to flakiness in msal-react handling after redirect.
// See github issue for details https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4658
function GraphAccessAuthEndHandlerImpl() {
  const {context} = useMicrosoftTeams();

  const handleRedirectPromiseLoader = usePromiseLoader<AuthenticationResult | null>();

  const instance = useInstanceVariable(() => {
    const msalConfig = getMsalConfig(`${config.host}/msteams/graphAuthEnd`);
    return new PublicClientApplication(msalConfig);
  });

  const [error, setError] = useState('');
  const [accessToken, setAccessToken] = useState('');

  useEffect(() => {
    if (!context) {
      return;
    }

    if (error) {
      microsoftTeams.authentication.notifyFailure(error);
      return;
    }

    if (accessToken) {
      microsoftTeams.authentication.notifySuccess(accessToken);
    }
  }, [accessToken, context, error]);

  useEffect(() => {
    handleRedirectPromiseLoader.await(
      instance.handleRedirectPromise(),
      authResult => setAccessToken(ensureExists(authResult).accessToken),
      err => setError(err.message),
    );
  }, [handleRedirectPromiseLoader, instance]);

  return <LoadingIndicator className={LoadingIndicator.style.medium} />;
}
export const GraphAccessAuthEndHandler = functionalComponent(
  'GraphAccessAuthEndHandler',
  GraphAccessAuthEndHandlerImpl,
);
```

Previous code using `msal-react` (would see previously reported issue `state_not_found: State not found: Cached State` transiently):
```
function GraphAccessAuthEndHandlerComponentImpl() {
  const {context} = useMicrosoftTeams();
  const {instance} = useMsal();
  const [error, setError] = useState('');
  const [accessToken, setAccessToken] = useState('');

  useEffect(() => {
    if (!context) {
      return;
    }
    if (error) {
      microsoftTeams.authentication.notifyFailure(error);
      return;
    }
    if (accessToken) {
      microsoftTeams.authentication.notifySuccess(accessToken);
    }
  }, [accessToken, context, error]);

  useEffect(() => {
    const callbackId = instance.addEventCallback(
      (message: {eventType: EventType; error: Error; payload: {accessToken: string}}) => {
        switch (message.eventType) {
          case EventType.LOGIN_FAILURE:
          case EventType.ACQUIRE_TOKEN_FAILURE:
            setError(message.error.message);
            break;
          case EventType.LOGIN_SUCCESS:
          case EventType.ACQUIRE_TOKEN_SUCCESS:
            setAccessToken(message.payload.accessToken);
            break;
        }
      },
    );

    return () => {
      if (callbackId) {
        instance.removeEventCallback(callbackId);
      }
    };
  }, [instance]);

  return <LoadingIndicator className={LoadingIndicator.style.medium} />;
}
const GraphAccessAuthEndHandlerComponent = functionalComponent(
  'GraphAccessAuthEndHandlerComponent',
  GraphAccessAuthEndHandlerComponentImpl,
);

function GraphAccessAuthEndHandlerImpl() {
  return (
    <MicrosoftTeamsMsalProvider redirectUri={`${config.host}/msteams/graphAuthEnd`}>
      <GraphAccessAuthEndHandlerComponent />
    </MicrosoftTeamsMsalProvider>
  );
}
```

**Comment by jasonnutter** [comment]
@benlieber-codaio Glad to hear you got it working. I'll see if I can figure out why your previous approach wasn't reliable.
